src/frontend/actors_generation_phase.nit

```diff
@@ -189,12 +186,46 @@ redef class Proxy{{{classname}}}
 end
 """)
 
-		redef_classes.add(
+	if mclassdef.is_intro then redef_classes.add(
 """
 redef class {{{classname}}}
-redef var async is lazy do return new Proxy{{{classname}}}.proxy(self)
+	var m = new Mutex
+	var lazy_proxy: Proxy{{{classname}}} is lazy do return new Proxy{{{classname}}}.proxy(self)
+
+	redef fun async: Proxy{{{classname}}} do
+		m.lock
+		var p = lazy_proxy
+		m.unlock
+		return p
+	end
 end
 """)
+
+	end
+
+	redef fun process_nmodule(nmodule) do
+		var mmod = nmodule.mmodule
+		if mmod == null then return
+
+		if generated_actor_modules.has(mmod.name) then return
+
+		var mclasses_defs = mmod.mclassdefs
+		for mclass_def in mclasses_defs do
+			var mclass = mclass_def.mclass
+			var actor = mclass.actor
+			if actor != null then generate_actor_classes(mclass_def, mmod)
+		end
+
+	end
+
+	redef fun process_annotated_node(nclass, nat)
+	do
+		if nat.n_atid.n_id.text != "actor" then return
+
+		if not nclass isa AStdClassdef then
```

Comment:
Because you add attributes to the annotated classes you should also check that they are either a concrete class or an abstract class, and not an interface, extern class or universal. This would be automatically detected if there was a common abstract class to all actored.